src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -0,0 +1,282 @@
+    private readonly Lazy<IProjectHotReloadNotificationService> _projectHotReloadNotificationService;
+    private readonly IProjectHotReloadAgent _hotReloadAgent;
+    private readonly IProjectThreadingService _threadingService;
+
+    // Protect the state from concurrent access. For example, our Process.Exited event
+    // handler may run on one thread while we're still setting up the session on
+    // another. To ensure consistent and proper behavior we need to serialize access.
+    private readonly ReentrantSemaphore _semaphore;
+
+    private readonly List<HotReloadSessionState> _activeSessionStates = [];
+    private HotReloadSessionState? _pendingSessionState = null;
+    private int _nextUniqueId = 1;
+    public bool HasActiveHotReloadSessions => _activeSessionStates.Count != 0;
+
+    [ImportingConstructor]
+    public ProjectHotReloadSessionManager(
+        ConfiguredProject project,
+        IProjectThreadingService threadingService,
+        Lazy<IHotReloadDiagnosticOutputService> hotReloadDiagnosticOutputService,
+        Lazy<IProjectHotReloadNotificationService> projectHotReloadNotificationService,
+        IProjectHotReloadLaunchProvider launchProvider,
+        IProjectHotReloadBuildManager buildManager,
+        IProjectHotReloadAgent hotReloadAgent)
+        : base(threadingService.JoinableTaskContext)
+    {
+        _configuredProject = project;
+        _unconfiguredProject = project.UnconfiguredProject;
+        _threadingService = threadingService;
+        _hotReloadDiagnosticOutputService = hotReloadDiagnosticOutputService;
+        _projectHotReloadNotificationService = projectHotReloadNotificationService;
+        _hotReloadAgent = hotReloadAgent;
+        _semaphore = ReentrantSemaphore.Create(
+            initialCount: 1,
+            joinableTaskContext: threadingService.JoinableTaskContext.Context,
+            mode: ReentrantSemaphore.ReentrancyMode.Freeform);
+    }
+
+    public Task<bool> TryCreatePendingSessionAsync(
+        IProjectHotReloadLaunchProvider launchProvider,
+        IDictionary<string, string> environmentVariables,
+        DebugLaunchOptions launchOptions,
+        ILaunchProfile launchProfile)
+    {
+        return _semaphore.ExecuteAsync(TryCreatePendingSessionInternalAsync).AsTask();
+
+        async ValueTask<bool> TryCreatePendingSessionInternalAsync()
+        {
+            if (await ProjectSupportsHotReloadAsync())
+            {
+                if (await ProjectSupportsStartupHooksAsync())
+                {
+                    string name = Path.GetFileNameWithoutExtension(_unconfiguredProject.FullPath);
+                    HotReloadSessionState hotReloadSessionState = new((HotReloadSessionState sessionState) =>
+                    {
+                        int count;
+                        lock (_activeSessionStates)
+                        {
+                            _activeSessionStates.Remove(sessionState);
```

Comment:
What if `Remove` returns `false`, indicating that the session wasn't actually in the set of those considered active? In such cases, it's possible that the hot reload state would be set to false multiple times.

Consider adding validation that the item was actually removed so that unexpected program flows are surfaced.


```suggestion
                            Assumes.False(_activeSessionStates.Remove(sessionState), "Cannot remove unknown hot reload session.");
```